### PR TITLE
Add nullable annotations to Mvc.Core/Infrastructure

### PR DESCRIPTION
**Pull request by pranavkm**

### CODE REVIEW

File: src/Mvc/Mvc.Core/src/Infrastructure/ActionSelector.cs [hunk 1]
```diff
@@ -121,17 +122,7 @@ public ActionDescriptor SelectBestCandidate(RouteContext context, IReadOnlyList<
...
 
-        /// <summary>
-        /// Returns the set of best matching actions.
-        /// </summary>
-        /// <param name="actions">The set of actions that satisfy all constraints.</param>
-        /// <returns>A list of the best matching actions.</returns>
-        private IReadOnlyList<ActionDescriptor> SelectBestActions(IReadOnlyList<ActionDescriptor> actions)
-        {
-            return actions;
-        }
```
Comment (JamesNK) [inline src/Mvc/Mvc.Core/src/Infrastructure/ActionSelector.cs:132]: In communist ASP.NET Core, every action is best action.

File: src/Mvc/Mvc.Core/src/Infrastructure/IActionResultExecutor.cs [hunk 2]
```diff
@@ -16,7 +18,7 @@ namespace Microsoft.AspNetCore.Mvc.Infrastructure
     /// <see cref="IActionResult.ExecuteResultAsync(ActionContext)"/> method of the corresponding action result type.
     /// Implementations should be registered as singleton services.
     /// </remarks>
-    public interface IActionResultExecutor<in TResult> where TResult : IActionResult
+    public interface IActionResultExecutor<in TResult> where TResult : notnull, IActionResult
```
Comment (JamesNK) [inline src/Mvc/Mvc.Core/src/Infrastructure/IActionResultExecutor.cs:21]: Is notnull a breaking change constraint?

File: src/Mvc/Mvc.Core/src/PublicAPI.Shipped.txt [hunk 3]
```diff
@@ -1357,57 +1357,57 @@ virtual Microsoft.AspNetCore.Mvc.ModelBinding.Validation.ValidationVisitor.Visit
...
-~Microsoft.AspNetCore.Mvc.Infrastructure.IActionSelector.SelectCandidates(Microsoft.AspNetCore.Routing.RouteContext context) -> System.Collections.Generic.IReadOnlyList<Microsoft.AspNetCore.Mvc.Abstractions.ActionDescriptor>
-~Microsoft.AspNetCore.Mvc.Infrastructure.IClientErrorFactory.GetClientError(Microsoft.AspNetCore.Mvc.ActionContext actionContext, Microsoft.AspNetCore.Mvc.Infrastructure.IClientErrorActionResult clientError) -> Microsoft.AspNetCore.Mvc.IActionResult
+Microsoft.AspNetCore.Mvc.Infrastructure.ContentResultExecutor.ContentResultExecutor(Microsoft.Extensions.Logging.ILogger<Microsoft.AspNetCore.Mvc.Infrastructure.ContentResultExecutor!>! logger, Microsoft.AspNetCore.Mvc.Infrastructure.IHttpResponseStreamWriterFactory! httpResponseStreamWriterFactory) -> void
+~Microsoft.AspNetCore.Mvc.Infrastructure.DefaultOutputFormatterSelector.DefaultOutputFormatterSelector(Microsoft.Extensions.Options.IOptions<Microsoft.AspNetCore.Mvc.MvcOptions!>! options, Microsoft.Extensions.Logging.ILoggerFactory! loggerFactory) -> void
```
Comment (JamesNK) [inline src/Mvc/Mvc.Core/src/PublicAPI.Shipped.txt:1372]: Do you know why `~` is kept on constructors after enabling nullable?

**Review by JamesNK — APPROVED** [review]

File: src/Mvc/Mvc.Core/src/Infrastructure/IActionResultExecutor.cs (hunk 2 above)
Comment (pranavkm) [reply src/Mvc/Mvc.Core/src/Infrastructure/IActionResultExecutor.cs:21]: It's a source breaking change in some cases. However in this particular case, MVC does not invoke this API when the action result is null, so it's kinda moot.

File: src/Mvc/Mvc.Core/src/PublicAPI.Shipped.txt (hunk 3 above)
Comment (pranavkm) [reply src/Mvc/Mvc.Core/src/PublicAPI.Shipped.txt:1372]: Not sure. I tried regening it and it didn't change the outcome. 🤷🏽

**Comment by AndrewMSHowe** [comment]
Hello, sorry I don't work for Microsoft. I am just trying to build aspnetcore. Not entirely sure why but I just went down a rabbit hole :) Anyway it doesn't build, and this change seems to be why. You made IActionInvokerFactory.CreateInvoker(ActionContext) return a nullable IActionInvoker, and now in ActionEndpointFactory.CreateRequestDelegate there's a line 'return invoker.InvokeAsync();' which gives CS8602 Dereference of a possibly null reference. Which is fair enough!

**Comment by JamesNK** [comment]
@AndrewMSHowe https://github.com/dotnet/aspnetcore/pull/28941

**Pull request merged**